unixfs/hamt/hamt.go

```diff
@@ -284,33 +284,124 @@ func (ds *Shard) getChild(ctx context.Context, i int) (child, error) {
 // loadChild reads the i'th child node of this shard from disk and returns it
 // as a 'child' interface
 func (ds *Shard) loadChild(ctx context.Context, i int) (child, error) {
+	c, err := ds.loadChildValue(i)
+	if err != nil {
+		return nil, err
+	}
+	if c != nil {
+		return c, nil
+	}
+
+	lnk := ds.nd.Links()[i]
+	nd, err := lnk.GetNode(ctx, ds.dserv)
+	if err != nil {
+		return nil, err
+	}
+	c, err = NewHamtFromDag(ds.dserv, nd)
+	if err != nil {
+		return nil, err
+	}
+
+	ds.children[i] = c
+	return c, nil
+}
+
+func (ds *Shard) missingChildShards() []*cid.Cid {
+	if len(ds.children) != len(ds.nd.Links()) {
+		panic("inconsistent lengths between children array and Links array")
```

Comment:
Are there any scenarios that this could happen with a maliciously crafted node?

Reply:
Not sure.  The reason I panic here is because I am unsure what to do with the error.

Reply:
Actually the reason for the panic is because it should not happen.  This is only called inside `preloadChildren` which already verifies the link.

Nevertheless I changed this method to return an error and handled the error by panicking in the fetcher code.

Reply:
Okay.  This should be fixed now.